File: java/leetcode/all-paths-from-source-to-target.java
class Solution {
    public static ArrayList<List<Integer>> result;
    private static ArrayList<Integer> path;

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        result = new ArrayList();
        path = new ArrayList();
        path.add(0);
        search(0, graph);
        return result;
    }
    
    private void search(int node, int[][] graph) {
        if (node == graph.length - 1) {
            result.add(new ArrayList<Integer>(path));
        } else {
            for (int i = 0; i < graph[node].length; i++) {
                int nextNode = graph[node][i];
                path.add(nextNode);
                search(nextNode, graph);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: javascript/leetcode/all-paths-from-source-to-target.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function(graph) {
    const result = [];
    const path = [0];
    const search = function (node, path) {
        if (node === graph.length - 1) {
            result.push(path.slice(0));
        } else {
            for (var i = 0; i < graph[node].length; i++) {
                path.push(graph[node][i]);
                search(graph[node][i], path);
                path.pop();
            }
        }
    }
    search(0, path);
    return result;
};
